server/schedule/checker/priority_checker.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"time"
+
+	"github.com/tikv/pd/server/schedule/placement"
+
+	"github.com/tikv/pd/pkg/cache"
+	"github.com/tikv/pd/server/config"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule/opt"
+)
+
+// the default value of priority queue size
+const defaultPriorityQueueSize = 1280
+
+// PriorityChecker ensures high priority region should run first
+type PriorityChecker struct {
+	cluster opt.Cluster
+	opts    *config.PersistOptions
+	queue   *cache.PriorityQueue
+}
+
+// NewPriorityChecker creates a priority checker.
+func NewPriorityChecker(cluster opt.Cluster) *PriorityChecker {
+	return &PriorityChecker{
+		cluster: cluster,
+		opts:    cluster.GetOpts(),
+		queue:   cache.NewPriorityQueue(defaultPriorityQueueSize),
+	}
+}
+
+// GetType returns PriorityChecker's type
+func (p *PriorityChecker) GetType() string {
+	return "priority-checker"
+}
+
+// RegionPriorityEntry records region priority info
+type RegionPriorityEntry struct {
+	Retry    int
+	Last     time.Time
+	regionID uint64
+}
+
+// ID implement PriorityQueueItem interface
+func (r RegionPriorityEntry) ID() uint64 {
+	return r.regionID
+}
+
+// NewRegionEntry construct of region priority entry
+func NewRegionEntry(regionID uint64) *RegionPriorityEntry {
+	return &RegionPriorityEntry{regionID: regionID, Last: time.Now(), Retry: 1}
+}
+
+// Check check region's replicas, it will put into priority queue if the region lack of replicas.
+func (p *PriorityChecker) Check(region *core.RegionInfo) {
+	makeupCount := 0
+	if p.opts.IsPlacementRulesEnabled() {
+		makeupCount = p.checkRegionInPlacementRule(region)
+	} else {
+		makeupCount = p.checkRegionInReplica(region)
+	}
+	priority := 0 - makeupCount
+	p.addPriorityQueue(priority, region.GetID())
```

Comment:
For those regions which don't need to make it up, do we still need to do it?

Reply:
this pr solve the recovery situation . it will speed the progress if some store broken and the region should make up replica after recovery by manual

Reply:
According to your implementation, each region, no matter if it should make up a replica, will enter addPriorityQueue, right?

Reply:
yes

Reply:
It's better to add it only make up the replica.

Reply:
the addPriorityQueue function will remove the region if it don't need to make up replica, otherwise, it will add the region to priority queue.